File: src/cpp/session/modules/prefs/UserState.cpp
/*
 * UserState.cpp
 *
 * Copyright (C) 2009-19 by RStudio, Inc.
 *
 * Unless you have received this program directly from RStudio pursuant
 * to the terms of a commercial license agreement with RStudio, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */

#include <core/json/JsonRpc.hpp>

#include <core/Exec.hpp>

#include <session/SessionOptions.hpp>
#include <session/SessionModuleContext.hpp>

#include "UserStateValues.hpp"
#include "UserStateDefaultLayer.hpp"
#include "UserStateLayer.hpp"
#include "Preferences.hpp"

using namespace rstudio::core;

namespace rstudio {
namespace session {
namespace modules {
namespace prefs {
namespace {

class UserState: public UserStateValues
{
   Error createLayers()
   {
      layers_.push_back(boost::make_shared<UserStateDefaultLayer>());  // STATE_LAYER_DEFAULT
      layers_.push_back(boost::make_shared<UserStateLayer>());         // STATE_LAYER_USER
      return Success();
   }
} s_state;

Error setState(const json::JsonRpcRequest& request,
               json::JsonRpcResponse* pResponse)
{
   json::Value val;
   Error error = json::readParams(request.params, &val);
   if (error)
      return error;

   s_state.writeLayer(STATE_LAYER_USER, val.get_obj()); 

   return Success();
}
} // anonymous namespace

json::Array userState()
{
   return s_state.allLayers();
}

Error initializeState()
{
   using namespace module_context;

   ExecBlock initBlock;
   initBlock.addFunctions()
      (bind(registerRpcMethod, "set_state", setState));
   Error error = initBlock.execute();
   if (error)
      return error;

   error = s_state.initialize();
   if (error)
      return error;

   return Success();
}

} // namespace state
} // namespace modules
} // namespace session
} // namespace rstudio

File: src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UserState.java
/*
 * UserState.java
 *
 * Copyright (C) 2020 by RStudio, PBC
 *
 * Unless you have received this program directly from RStudio pursuant
 * to the terms of a commercial license agreement with RStudio, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */
package org.rstudio.studio.client.workbench.prefs.model;

import com.google.gwt.core.client.JsArray;
import com.google.inject.Inject;
import com.google.inject.Singleton;

import org.rstudio.core.client.CommandWithArg;
import org.rstudio.core.client.Debug;
import org.rstudio.studio.client.RStudioGinjector;
import org.rstudio.studio.client.application.events.EventBus;
import org.rstudio.studio.client.common.satellite.Satellite;
import org.rstudio.studio.client.common.satellite.SatelliteManager;
import org.rstudio.studio.client.server.ServerError;
import org.rstudio.studio.client.server.ServerRequestCallback;
import org.rstudio.studio.client.server.Void;
import org.rstudio.studio.client.workbench.model.Session;
import org.rstudio.studio.client.workbench.prefs.events.UserStateChangedEvent;

@Singleton
public class UserState extends UserStateAccessor implements UserStateChangedEvent.Handler
{
   @Inject
   public UserState(Session session, 
                  EventBus eventBus,
                  PrefsServerOperations server,
                  SatelliteManager satelliteManager)
   {
      super(session.getSessionInfo(),
            (session.getSessionInfo() == null ? 
               JsArray.createArray().cast() :
               session.getSessionInfo().getUserState()));

      session_ = session;
      server_ = server;
      satelliteManager_ = satelliteManager;

      eventBus.addHandler(UserStateChangedEvent.TYPE, this);
   }
   
   public void writeState()
   {
      writeState(null);
   }

   public void writeState(CommandWithArg<Boolean> onCompleted)
   {
      updatePrefs(session_.getSessionInfo().getUserState());
      server_.setUserState(
         session_.getSessionInfo().getUserStateLayer().getValues(),
         new ServerRequestCallback<Void>() 
         {
            @Override
            public void onResponseReceived(Void v)
            {
               UserStateChangedEvent event = new UserStateChangedEvent(
                              session_.getSessionInfo().getUserStateLayer());

               if (Satellite.isCurrentWindowSatellite())
               {
                  RStudioGinjector.INSTANCE.getEventBus()
                     .fireEventToMainWindow(event);
               }
               else
               {
                  // let satellites know prefs have changed
                  satelliteManager_.dispatchCrossWindowEvent(event);
               }
               if (onCompleted != null)
               {
                  onCompleted.execute(true);
               }
            }
            @Override
            public void onError(ServerError error)
            {
               Debug.logError(error);
            }
         });
   }
   
   @Override
   public void onUserStateChanged(UserStateChangedEvent e)
   {
      syncPrefs(e.getName(), e.getValues());
   }

   private final Session session_;
   private final PrefsServerOperations server_;
   private final SatelliteManager satelliteManager_;
}
